Added favicon and publisher name to API

BScong opened the pull request:
See comments in issue #65

dylano commented on command/CommandController.js:
```diff
@@ -264,15 +274,44 @@ function getArticleMetadata(pocketArticle) {
-    lengthMinutes,
-    imageURL: pocketArticle.top_image_url
-  };
+  let faviconPromise = getFavicons(pocketArticle.resolved_url);
```
Let's protect this section with a request parameter will allow the Alexa skill to skip this additional metadata. We won't be using it there and the audio experience is more sensitive to long response times.

BScong replied:
Should I check the user-agent from the API call to decide if we should skip the favicon and publisher name calls?

dylano replied:
I don't know if we want to make assumptions based on user agent. I would lean toward something more explicit to give the client control over what they want to get back, so add a new body param like `completeData=1` or `extendedData=1` to trigger these additional queries.

BScong commented on command/CommandController.js:
```diff
@@ -264,15 +274,44 @@ function getArticleMetadata(pocketArticle) {
+    } else {
+      let a = document.createElement('a');
+      a.href = pocketArticle.resolved_url;
+      publisherName = a.hostname;
```
Done

BScong commented:
I can't reproduce the 400 error when having 50+ articles but the API call time is becoming very long with this number of articles, maybe we should paginate this if used on mobile?

BScong commented:
The npm-metadata call can also return favicons if provided by the website. Happens on ~90% of news websites I tested. So I guess we can only make that one call? But then if it doesn't provide the icon, I have to make the second call using the get-website-favicon package. So that will take longer because it won't be parallelized , but only happens ~10% of the time

BScong commented:
I still think the best way to do it is to store in a database the websites that we already requested so we don't have to fetch metadata each time... If an user has 10 articles from the same website, we are currently making the same calls 10 times.

BScong commented:
Updated to use database (DynamoDB) to store favicons and domain names (expires after 30 days).  Also added extendedData request parameter.
Measured times : https://docs.google.com/spreadsheets/d/13KfeV25gXfd-Nu23TKlYlu1QZ8SpFLrtXYotJuS-qck/edit?usp=sharing

dylano commented on command/CommandController.js:
```diff
@@ -65,15 +67,16 @@ router.post('/intent', VerifyToken, async function(req, res) {
     switch (req.body.cmd) {
       case 'ScoutTitles':
-        scoutTitles(getBody, res);
+        scoutTitles(getBody, res, req.body.extendedData == true);
```
👍

dylano commented on command/CommandController.js:
```diff
@@ -291,23 +317,56 @@ function getArticleMetadata(pocketArticle) {
+      author,
+      lengthMinutes,
+      imageURL: pocketArticle.top_image_url
+    };
```
Definitely want to clean up these repeated object structures. I've also been favoring async/await over vanilla promises when we can. An approach like the following reads better, where you declare the object once and then add in the extended properties as appropriate:
```
  const result = {
    item_id: pocketArticle.item_id,
    sort_id: pocketArticle.sort_id,
    resolved_url: pocketArticle.resolved_url,
    title: pocketArticle.resolved_title,
    author,
    lengthMinutes,
    imageURL: pocketArticle.top_image_url
  };

  if (extendedData) {
    try {
      const faviconData = await faviconHelper.getWebsiteFavicon(
        pocketArticle.resolved_url
      );
      result.publisher = faviconData.website_name;
      result.icon_url = faviconData.favicon_url;
    } catch (err) {
      result.publisher = faviconHelper.getHostname(pocketArticle.resolved_url);
      result.icon_url = '';
    }
  }
  return result;
```

BScong replied:
Done

dylano commented on command/CommandController.js:
```diff
@@ -238,27 +253,38 @@ function scoutSummaries(getOptions, jsonBodyAttr, urlAttr, titleAttr, res) {
-          return a.sort_id - b.sort_id;
-        });
+        Promise.all(articlesPromises).then(function(values) {
+          faviconHelper.clearCurrentRequests();
```
This call to faviconHelper here feels really out of context to the other things that are happening. From reading through the rest of this function, you would have no idea what this call was for or that you were expected to do some cleanup after calling `getArticleMetadata()`. I think the currentRequests activity needs to be handled inside of the FaviconHelper file.

BScong replied:
See comment on currentRequests[hostname]

dylano commented:
Overall functionality is great 🥇  
This new branch was able to handle the 50 articles in my large test account. From the logs I can see there were a couple of errors but it didn't kill the request this time:
```
...
getWebsiteFavicon for medium.com
getWebsiteFavicon for codeburst.io
Error occured during Favicon Promise calls.
Error occured during Favicon Promise calls.
storeWebsiteFavicon for andrewhavens.com: , andrewhavens.com
storeWebsiteFavicon for labs.chiedo.com: https://labs.chiedo.com/wp-content/themes/chiedolabs/favicon.ico?v=2, Chiedo Labs
...
```

BScong commented on command/FaviconHelper.js:
```diff
@@ -0,0 +1,86 @@
+      let self = this;
+      if (!self.currentRequests[hostname]) {
+        let faviconPromise = getFavicons(link);
+        let metadataPromise = scrape(link);
```
So I should make two separate tables (and two helpers FaviconHelper, PublisherHelper) : one linking the hostname to the favicon and one linking the hostname to the publisher name?

dylano replied:
No, it makes sense to have them in one table. I would just make this function more flexible so that you can retrieve either or both types of data depending on what client is asking for. The AudioFileLocations data does something similar for handling the full-audio & summary files.

BScong replied:
Got it

BScong replied:
Done, parameter type to specify between 'all', 'favicon' and 'publisher'. Defaults to 'all'.
Moved the fetching to another function, also depending on type and updated currentRequests var to reflect this

BScong commented on command/FaviconHelper.js:
```diff
@@ -0,0 +1,86 @@
+            console.log('Error occured during Favicon Promise calls.');
+          });
+      }
+      return self.currentRequests[hostname];
```
So this is a little bit tricky. The request is complete but since other calls can be made to the same hostname, I should keep it so I don't make the same call twice (for example when title 1 is on website A and title 49 is also on website A). This is also linked to your comment on the clearCurrentRequests() call

BScong replied:
But I agree with you that the clearCurrentRequests() call is out of context and that everything about the currentRequests things should be in the Helper. Will try to come up with a solution

BScong replied:
One solution would be to remove it from currentRequest when the call to database.storeWebsiteFavicon has ended. But I'm little afraid of the concurrent executions there... If at the same time we have a database.getWebsiteFavicon and a database.storeWebsiteFavicon calls, we could have a case where getWebsiteFavicon from the Helper returns undefined.
I think the best way is to clean currentRequests when all the requests are done to avoid those kind of problems, that's why I put the cleaning in a Promise.all call, but I agree with you that it's quite a "dirty" way.

BScong commented:
Could you send me the articles causing those errors? I am unable to reproduce those on my side with my articles

dylano commented:
Errors:
```
Error occured during Favicon Promise calls: 403 @ code.tutsplus.com
Error occured during Favicon Promise calls: 400 @ cgpgrey.com
```

* https://code.tutsplus.com/tutorials/introduction-to-css-grid-layout-with-examples--cms-25392
* http://www.cgpgrey.com/blog/voyage-to-nowhere

BScong commented:
Investigated the errors.
Comes from the metadata calls.
One isn't supported (400) and one forbids calls (403) and wants us to solve a captcha. 

Modified to catch the errors individually depending on the call, so that we still have a database entry even if one of the calls failed.

The problem is that we will re-call them each time we see the domain because we don't have data on them in the database, and each time the call will fail (until they support it).

dylano commented:
How about storing a value there which indicates that it is unavailable (but still pass back the empty value to the client). That way we can avoid querying them repeatedly for it (at least until the record expires).

BScong commented:
Nice idea. I now update the timestamp when I get errors, so that we don't resend same requests.
